cli/server.go

```diff
@@ -1376,6 +1106,172 @@ func newProvisionerDaemon(
 	}), nil
 }
 
+// CommonServerCmd is the common elements of starting a server daemon.
+// This can be used by both workspace proxies and the primary coderd.
+type CommonServerCmd struct {
+	Ctx       context.Context
+	NotifyCtx context.Context
+
+	HTTPServers *HTTPServers
+	HTTPClient  *http.Client
+	LocalURL    *url.URL
+
+	Logger             slog.Logger
+	PrometheusRegistry *prometheus.Registry
+	Tracer             trace.TracerProvider
+	// SQLDriver is the driver that the tracer is active on.
+	SQLDriver string
+
+	AppHostname      string
+	AppHostnameRegex *regexp.Regexp
+	RealIPConfig     *httpmw.RealIPConfig
+
+	closeFuncs []func()
+}
+
+// Close closes all the resources associated with the CommonServerCmd.
+func (c *CommonServerCmd) Close() {
+	for _, f := range c.closeFuncs {
+		f()
+	}
+}
+
+func (c *CommonServerCmd) addClose(f func()) {
+	c.closeFuncs = append(c.closeFuncs, f)
+}
+
+// SetupServerCmd sets up the common elements of starting a server daemon.
+// This is used by both coderd and workspace proxies.
+func SetupServerCmd(inv *clibase.Invocation, cfg *codersdk.DeploymentValues) (_ *CommonServerCmd, err error) {
```

Comment:
I think this is too broad of a function name. Is there a reason this needs to be abstracted together and can't just be duplicated? I'm having a difficult time understanding why I'd put something here, and whether it would break workspace proxies or not.

Reply:
I can duplicate it, and maybe that's the way to go. I was just trying not to duplicate all this logic across the Coderd and workspace proxy. 

I'm just afraid of regressions when things are not properly done in both places.

----

My other idea was to refactor this into as many functions as I can so at least the shared code is a bit less LOC

Reply:
If we don't share the shared code we will start seeing devs do something in server.go and not do it in proxy.go just like we did in V1. Code duplication is off the table IMO because of this.

Just using functions and not a big "CommonServerCmd" struct would also work

Reply:
I can whip up the making common functions option. It's not as nice since I have to export a lot of things from this cli package, and new things still have be copied into 2 locations.